quickwit/quickwit-indexing/src/source/queue_sources/local_state.rs

```diff
@@ -0,0 +1,150 @@
+// Copyright (C) 2024 Quickwit, Inc.
+//
+// Quickwit is offered under the AGPL v3.0 and as commercial software.
+// For commercial licensing, contact us at hello@quickwit.io.
+//
+// AGPL:
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+use std::collections::{BTreeMap, BTreeSet, VecDeque};
+
+use quickwit_metastore::checkpoint::PartitionId;
+
+use super::message::{CheckpointedMessage, InProgressMessage, PreProcessedMessage};
+use super::visibility::VisibilityTaskHandle;
+use super::Categorized;
+
+/// Tracks the state of the the queue messages that is known to the owning
+/// indexing pipeline.
+///
+/// Messages first land in the `ready` queue. In parallel, they are also
+/// recorded as `in_flight`, where the visibility handle is kept alive. Messages
+/// are then moved to `in_progress` to track the reader's progress. It's only
+/// when the events are published that the in-flight visiblity handles are
+/// dropped and the partition is marked as `completed``.
+#[derive(Default)]
+pub struct QueueLocalState {
+    in_flight: BTreeMap<PartitionId, VisibilityTaskHandle>,
+    ready: VecDeque<CheckpointedMessage>,
+    in_progress: Option<InProgressMessage>,
+    completed: BTreeSet<PartitionId>,
```

Comment:
we need comment explaining what all of these are about...
the naming is incredibly confusing.

I suspect better names will emerge from commenting.

Reply:
I added some comments but better names are still to be found

Reply:
I made the names significantly more explicit.